modules/password/password.go

```diff
@@ -0,0 +1,74 @@
+// Copyright 2019 The Gitea Authors. All rights reserved.
+// Use of this source code is governed by a MIT-style
+// license that can be found in the LICENSE file.
+
+package password
+
+import (
+	"crypto/rand"
+	"math/big"
+	"regexp"
+	"sync"
+
+	"code.gitea.io/gitea/modules/setting"
+)
+
+var matchComplexities = map[string]regexp.Regexp{}
+var matchComplexityOnce sync.Once
+var validChars string
+
+var validComplexities = map[string]string{
+	"lower": "abcdefghijklmnopqrstuvwxyz",
+	"upper": "ABCDEFGHIJKLMNOPQRSTUVWXYZ",
+	"digit": "0123456789",
+	"spec":  "_-",
```

Comment:
Special character list should be increased with additional characters as recommended by OWASP (https://www.owasp.org/index.php/Password_special_characters)
```suggestion
	"spec":  " !\"#\\$%&'\\(\\)\\*\\+,-\\./:;<=>\\?@\\[\\\\\\]\\^_\x60{\\|}~",
```
Hope I got escaping right :)

Reply:
Oh, I'm sorry! It was me that recommended reducing the set in order to avoid problems with several parsers (like shells). It got reduced too much, though. I'd certainly remove the space from the list, as it can be confusing if it were to appear at the beginning or the end of the string.

To avoid dealing with backslashes at the golang level, the string can simply be enclosed with backticks:
```
"spec": ` !"#$%&'()*+,-./:;<=>?@[\\]^_{|}~` + "`"
```

For a regular expression with a range/list of characters (`[abc123]`), you only need to escape the characters that have special meaning inside the range, namely: `\`, `]`, `-`, `^`. But even most of that you can avoid by: placing `]` first, right after the `[`; `-` last and `^` anywhere but first:

```
"spec" := `][ !"#$%&'()*+,./:;<=>?@\\^_{|}~` + "`-"
"specregexstr  := "[" + spec + "]"
```

Reply:
Escaping would be better

Reply:
When generating it could be checked if space is in the begging or end but I don't think we should because of that limit user to not count space as special character

Reply:
Increased the number of special characters.